code/modules/mob/living/carbon/human/species_types/ghost.dm

```diff
@@ -0,0 +1,182 @@
+	desc = "Toggles your ability to phase through everything, including your gear and any incompatible organs/limbs."
+	button_icon = 'icons/hud/actions.dmi'
+	button_icon_state = "ghost"
+
+/datum/action/innate/toggle_passthrough/Grant(mob/grant_to)
+	. = ..()
+	RegisterSignal(grant_to, COMSIG_CARBON_POST_ATTACH_LIMB, PROC_REF(on_new_limb))
+
+/datum/action/innate/toggle_passthrough/Remove(mob/remove_from)
+	swap_mode(force_off = TRUE)
+	UnregisterSignal(remove_from, COMSIG_CARBON_POST_ATTACH_LIMB)
+	return ..()
+
+/datum/action/innate/toggle_passthrough/Activate()
+	. = ..()
+	if(!iscarbon(owner))
+		return
+	swap_mode()
+
+/datum/action/innate/toggle_passthrough/IsAvailable(feedback)
+	if(LAZYLEN(owner.reagents))
+		if(locate(/datum/reagent/water/holywater) in owner.reagents)
+			return FALSE
+	return ..()
+
+///Called when the owner of this action gets a new limb, if it isn't a ghost-limb
+///the action will turn itself off, and you'll lose said limb if you try using this action again.
+/datum/action/innate/toggle_passthrough/proc/on_new_limb(mob/source, obj/item/bodypart/new_part, special)
+	SIGNAL_HANDLER
+	if(!iscarbon(owner))
+		return
+	if(!(new_part.bodytype & BODYTYPE_GHOST))
+		swap_mode(force_off = TRUE)
+
+///Swaps the mode, allowing us to phase through stuff but drops everything. Optional 'force_off' arg to prevent being able to turn it on.
+/datum/action/innate/toggle_passthrough/proc/swap_mode(force_off)
+	//we can only turn off, early return if we're trying to turn it on instead.
+	if(force_off && HAS_TRAIT(owner, TRAIT_NO_FLOATING_ANIM))
+		return
+
+	var/mob/living/carbon/carbon_owner = owner
+	var/datum/species/carbon_species = carbon_owner.dna.species
+
+	//drop any limbs & organs that can't phase, now that you're phasing.
+	for(var/obj/item/bodypart/bodypart as anything in carbon_owner.bodyparts)
+		if(!(bodypart.bodytype & BODYTYPE_GHOST))
+			bodypart.drop_limb(special = FALSE, dismembered = FALSE, move_to_floor = TRUE)
+	for(var/obj/item/organ/organ as anything in carbon_owner.organs)
+		if(!(organ.organ_flags & ORGAN_GHOST))
+			organ.Remove(owner, special = FALSE)
+
+	if(HAS_TRAIT(carbon_owner, TRAIT_NO_FLOATING_ANIM))
+		REMOVE_TRAIT(carbon_owner, TRAIT_NO_FLOATING_ANIM, SPECIES_TRAIT)
+		carbon_owner.add_traits(list(
+			TRAIT_MOVE_PHASING,
+			TRAIT_PIERCEIMMUNE,
+			TRAIT_HANDS_BLOCKED, //MOBILITY_USE | MOBILITY_PICKUP | MOBILITY_STORAGE
+			TRAIT_PULL_BLOCKED, //MOBILITY_PULL
+			TRAIT_UI_BLOCKED, //MOBILITY_UI
+			), SPECIES_TRAIT)
+		carbon_species.update_no_equip_flags(carbon_owner, ALL)
+		RegisterSignal(carbon_owner, COMSIG_MOB_CLIENT_PRE_LIVING_MOVE, PROC_REF(attempt_move))
+	else
+		ADD_TRAIT(carbon_owner, TRAIT_NO_FLOATING_ANIM, SPECIES_TRAIT)
+		carbon_owner.remove_traits(list(
+			TRAIT_MOVE_PHASING,
+			TRAIT_PIERCEIMMUNE,
+			TRAIT_HANDS_BLOCKED,
+			TRAIT_PULL_BLOCKED,
+			TRAIT_UI_BLOCKED,
+			), SPECIES_TRAIT)
+		carbon_owner.mobility_flags = initial(carbon_owner.mobility_flags)
```

Comment:
Pretty sure you shouldn't be messing with mobi flags directly

Reply:
oh oops forgot to remove it in https://github.com/tgstation/tgstation/pull/90711/commits/b632b5c272e0b0b443da6f51e1adefcbbd419afd